asio/include/asio/detail/macos_fenced_block.hpp

```diff
@@ -19,7 +19,40 @@
 
 #if defined(__MACH__) && defined(__APPLE__)
 
+#include <AvailabilityMacros.h>
+
+// Availability.h was introduced for supporting Mac OSX 10.6 and iOS
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= 1060 \
+  || defined(__IPHONE_OS_VERSION_MIN_REQUIRED)
+#include <Availability.h>
+#endif
+
+#if defined(__IPHONE_OS_VERSION_MIN_REQUIRED)
+
+// OSMemoryBarrier was deprecated in iOS 10
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 100000
+#define ASIO_APPLE_USE_ATOMIC_FENCE 1
+#endif
+
+// Use MAC_OS_X_VERSION_MIN_REQUIRED instead of __MAC_OS_X_VERSION_MIN_REQUIRED
+// to keep compatibility with AvailabilityMacros.h
+#elif defined(MAC_OS_X_VERSION_MIN_REQUIRED)
+
+// OSMemoryBarrier was deprecated in Mac OSX 10.12
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101200
+#define ASIO_APPLE_USE_ATOMIC_FENCE 1
```

Comment:
Should this macro name be something like 'ASIO_APPLE_USE_PROPRIETORY_FENCE' with flipped logic? So that standard code has a "default" flow feeling?